### DISCUSSION TITLE
Profiling of the DeepMD-kit

### DISCUSSION THREAD

**Discussion by AnuragKr** [deepmodeling/deepmd-kit#1842] [Q&A] [answered]
I want to do profiling of the deepmd-kit to understand the complete flow and which Cuda kernel taking more time or less.
The command I used - 
1. CUDA_VISIBLE_DEVICES=1 mpirun -l -launcher=fork -hosts=localhost -np 1 nsys profile -o deepmd_report --status=true dp train --mpi-log=workers input.json

![Screenshot 2022-08-02 at 4 01 38 PM](https://user-images.githubusercontent.com/10074710/182359657-0266578d-deff-456b-bda3-b3d0051414f6.png)

2. nsys profile -o deepmd_nsys_report --stats=true  python train.py
![Screenshot 2022-08-02 at 4 36 08 PM](https://user-images.githubusercontent.com/10074710/182360176-df4ba21e-8204-4a41-b105-c1fe7597bf39.png)

Another Issue I am encountering --
I am making changes locally to the code but I am unable to see any changes. I have installed via source code. 

@denghuilu  Please suggest where I am going wrong. I am using nsys-cli.

**Comment by denghuilu** [comment]
I think the execution time of the profiling progress was too long. You can control the number of training steps to be about 10 steps by modifying the parameters in the training script. In fact, the DP training of each step basically goes through the same network size, so a short-time profiling still works.

**Reply by AnuragKr** [reply]
I tried to do profiling using this command by reducing the training step to 100--
Command ---
CUDA_VISIBLE_DEVICES=0,1 mpirun -l -launcher=fork -hosts=localhost -np 1 nsys profile -o single_gpu_profile_100 -t cuda,mpi,nvtx  --delay 30 --duration 20 dp train --mpi-log=workers input.json
Output --
single_gpu_profile_100.nsys-rep

For GUI visualization I opened this output file using via Nsight System following output I got--
![Screenshot 2022-08-18 at 11 02 55 PM](https://user-images.githubusercontent.com/10074710/185458549-f58dea77-e6ef-4887-8ff5-05c491e5176a.png)


I want to know whether I am doing it correctly or What are the steps to do for analyzing via nsys.

**Reply by denghuilu** [reply]
Maybe you can profiling through the [nsys command line tool](https://docs.nvidia.com/nsight-systems/UserGuide/index.html#cli-profiling) and then view the relevant results locally. This can avoid some communication link problems.

**Reply by AnuragKr** [reply]
Hi @denghuilu ,

Thanks for your response to the problem mentioned above. Now I am trying to calculate FLOPS using these commands--
1. Command --  nv-nsight-cu-cli --devices 0 --query-metrics -o single_gpu_ice_triple_I dp train input.json
    Output -- 
![nsight_compute](https://user-images.githubusercontent.com/10074710/187828745-525ef6f6-0b71-4136-a383-6ce521e65770.png)

2. Command -- nvprof --metrics flops_sp --metrics flops_sp_add --metrics flops_sp_mul --metrics flops_sp_fma --metrics flops_sp_special dp train input.json
Output --
![Screenshot 2022-08-31 at 7 25 07 PM](https://user-images.githubusercontent.com/10074710/187828984-807c6cff-9776-4b1f-aedf-b5976303d5ed.png)

Please Suggest --
1. Which approach to use to calculate FLOP or Kernel level profiling.In research [paper](https://arxiv.org/abs/2005.00223) it is mentioned nvprof used.
2. For nvprof command anything I am doing it worong.

**Reply by AnuragKr** [reply]
@denghuilu  Please guide me in resolving the FLOP count issue --
Q1.) Did you come across this issue while doing profiling -- [link](https://developer.nvidia.com/nvidia-development-tools-solutions-err_nvgpuctrperm-permission-issue-performance-counters). Is it mandatory to give hardware access to do profiling?
Q2.) What metrics look out for evaluating performance? 
My profiling window looks like this --
![Screenshot 2022-12-17 at 6 50 57 PM](https://user-images.githubusercontent.com/10074710/208243995-3b491de9-a2df-4bf5-b6e4-f1da153bf627.png)

**Comment by AnuragKr** [comment, marked as answer]
Solution--
1. Issue -- ERR_NVGPUCTRPERM: Permission issue with Performance Counters 
     Steps to solve  (Ask System Administrator to follow these steps)--
     1.  Write 
        
         ```bash 
         options nvidia "NVreg_RestrictProfilingToAdminUsers=0 
         ``` 
         into /etc/modprobe.d/nvidia-prof.conf

      2.  ```bash 
            sudo update-initramfs -u -k all
           ```
      
     3. Reboot


2.  Kernel level profiling and FLOP count can be done now using Nsight compute rather than nvprof.
     ```bash
     CUDA_VISIBLE_DEVICES=1 ncu --target-processes all -o profile mpirun -np 1 dp train  input.json
      ```
   For more advanced profiling options - https://docs.nvidia.com/nsight-compute/NsightComputeCli/index.html
